code/datums/ai/generic_actions.dm

```diff
@@ -156,3 +156,26 @@
 
 	if(QDELETED(target) || prob(10)) // Even if we don't finish it all we can randomly decide to be done
 		finish_action(controller, TRUE)
+
+/**find and set
+ * Finds an item near themselves, sets a blackboard key as it. Very useful for ais that need to use machines or something.
+ * if you want to do something more complicated than find a single atom, change the search_tactic() proc
+ * cool tip: search_tactic() can set lists
+ */
+/datum/ai_behavior/find_and_set
+	action_cooldown = 5 SECONDS
+	//optional, don't use if you're changing search_tactic()
+	var/locate_path
+	var/bb_key_to_set
+
+/datum/ai_behavior/find_and_set/perform(delta_time, datum/ai_controller/controller)
+	. = ..()
+	var/find_this_thing = search_tactic(controller)
+	if(find_this_thing)
+		controller.blackboard[bb_key_to_set] = find_this_thing
+		finish_action(controller, TRUE)
+	else
+		finish_action(controller, FALSE)
+
+/datum/ai_behavior/find_and_set/proc/search_tactic(datum/ai_controller/controller)
+	return locate(locate_path) in oview(7, controller.pawn)
```

Comment:
I don't know much about our AI implementation, but that "7" should be configurable.

Also, I'm assuming this won't perform too badly because of the 5 second cooldown?

Reply:
Yes, but more importantly this AI is very rare, not long lasting, and not spammable. The AI kills itself if it hits a target, so.